### PR TITLE
CAN: Initialisation Timeout

### PR DESCRIPTION
**Pull request by garethpotter**
Add an (optional) timeout on CAN initialisation.

### CODE REVIEW

File: src/can.c [hunk 1]
```diff
@@ -120,6 +121,18 @@ static void thingset_can_addr_claim_tx_cb(const struct device *dev, int error, v
     }
 }
 
+static void thingset_can_addr_discover_tx_cb(const struct device *dev, int error, void *user_data)
+{
+    struct thingset_can *ts_can = user_data;
+
+    if (error == 0) {
+        k_event_post(&ts_can->events, EVENT_ADDRESS_DISCOVER_MSG_SENT);
```
Comment (martinjaeger) [inline src/can.c:129]: I'm not sure I understand correctly, but it looks like this new event is never waited for.

File: include/thingset/can.h [hunk 2]
```diff
@@ -358,11 +358,13 @@ int thingset_can_set_item_rx_callback_inst(struct thingset_can *ts_can,
...
  * @param can_dev Pointer to the CAN device that should be used.
  * @param bus_number Assigned bus number of this CAN device (ignored if bridge routing is used)
+ * @param timeout_ms Timeout in milliseconds to wait for initialisation. Pass -1 to wait
+ * indefinitely.
  *
  * @returns 0 for success or negative errno in case of error
  */
 int thingset_can_init_inst(struct thingset_can *ts_can, const struct device *can_dev,
-                           uint8_t bus_number);
+                           uint8_t bus_number, int timeout_ms);
```
[the remark this answers is not in the document]
Comment (garethpotter) [reply include/thingset/can.h:367]: That was my original intention, but I couldn't figure out how to do that at the time. I will have another think.

[the remark this answers is not in the document]
Comment (garethpotter) [reply include/thingset/can.h]: I have now belatedly reworked this to use `k_timeout_t`.

**Comment by garethpotter** [comment]
I've now finally had another crack at this, as you will see. I noticed as I was adding a new event that the existing `ALREADY_USED` event had a value of `0x03`; as these are flags, it should surely have been `BIT(3)`, so I have amended accordingly.

To preserve what looked like the existing behaviour (and what makes sense to me reading it), I now explicitly clear all three of the events, two of which were being implicitly cleared by the call to `k_event_clear` at line 663. I also pass the `thingset_can_addr_claim_tx_handler` callback to `can_send`, so that that callback can set the `MSG_SENT` event which is awaited at line 702. I think this all looks _correct_ to me now but I welcome your thoughts.

File: src/can.c [hunk 3]
```diff
@@ -586,6 +605,8 @@ int thingset_can_init_inst(struct thingset_can *ts_can, const struct device *can
 #endif
     k_sem_init(&ts_can->request_response.sem, 1, 1);
     k_sem_init(&ts_can->report_tx_sem, 0, 1);
+    k_timer_init(&ts_can->timeout_timer, thingset_can_timeout_timer_expired,
+                 thingset_can_timeout_timer_stopped);
```
Comment (martinjaeger) [inline src/can.c:609]: If the stop function is not needed, just set it to `NULL`:
```suggestion
    k_timer_init(&ts_can->timeout_timer, thingset_can_timeout_timer_expired, NULL);
```

File: src/can.c [hunk 4]
```diff
@@ -565,8 +573,19 @@ static void thingset_can_reqresp_sent_callback(int result, void *arg)
...
+{
+    struct thingset_can *ts_can = CONTAINER_OF(timer, struct thingset_can, timeout_timer);
+    k_event_set(&ts_can->events, EVENT_ADDRESS_CLAIM_TIMED_OUT);
+}
+
+static void thingset_can_timeout_timer_stopped(struct k_timer *timer)
+{
+    ARG_UNUSED(timer);
+}
+
```
Comment (martinjaeger) [inline src/can.c:586]: Not needed if we set the parameter to `NULL`.

**Review by martinjaeger — COMMENTED** [review]
Looks good, thanks for the update. Just got a very minor suggestion.

**Review by martinjaeger — APPROVED** [review]
Looks good, thanks for the update.

**Pull request merged**